### PR TITLE
Train NLVR parser with multi-world consistency

### PR DESCRIPTION
**Pull request by pdasigi**
Summary of changes:
1) Added a pre-processing script for grouping instances in the nlvr dataset that have the same sentence across multiple worlds
2) Changed the dataset reader to allow reading grouped datasets
3) Changed the model to take a list of worlds and a list of labels for each instance (even in the ungrouped case, in which case the lists will contain single elements) and redefined denotation loss as being 0 iff denotations in *all* worlds are correct
4) Defined a new metric called "consistency" which is 1 iff all the denotations are correct. Accuracy is the average number of denotations that are correct per instance. Note that if the input dataset is ungrouped, accuracy and consistency will be equal.

I'll rebase this after #1013 gets merged. This PR doesn't depend on that though.

### CODE REVIEW

File: allennlp/data/dataset_readers/nlvr.py [hunk 1]
```diff
@@ -26,9 +26,12 @@ class NlvrDatasetReader(DatasetReader):
     instances from text, this class contains a method for creating an agenda of actions that each
     sentence triggers.
 
-    We process the json version of the dataset (http://lic.nlp.cornell.edu/nlvr/) here, to read in
-    the structured representations of the synthetic images instead of dealing with the actual images
-    themselves. The format of each line in the jsonl file is
+    We process the either the original json version of the NLVR dataset
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/nlvr.py]: "the either the"

**Comment by DeNeutoy** [comment]
The name of this PR is deep. 🔮

File: allennlp/data/dataset_readers/nlvr.py [hunk 2]
```diff
@@ -26,9 +26,12 @@ class NlvrDatasetReader(DatasetReader):
...
+    (http://lic.nlp.cornell.edu/nlvr/) or the processed version where we group all the worlds that a
+    sentence appears in here.
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/nlvr.py]: You should mention the script that you use to process the data.

File: allennlp/data/dataset_readers/nlvr.py [hunk 3]
```diff
@@ -87,53 +97,67 @@ def _read(self, file_path: str):
...
         fields = {"sentence": sentence_field,
                   "agenda": agenda_field,
-                  "world": world_field,
+                  "worlds": worlds_field,
                   "actions": action_field}
-        if label:
-            label_field = LabelField(label, label_namespace='denotations')
-            fields["label"] = label_field
+        if labels:
+            labels_field = ListField([LabelField(label, label_namespace='denotations') for label in
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/nlvr.py]: It's easier to read if you wrap on the `for`.

File: scripts/nlvr/group_nlvr_worlds.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        # "identifier" in the original dataset looks something like 4055-3, where 4055 is common
+        # across all four instances with the same sentence, but different worlds, and the suffix
+        # differentiates among the four instances.
+        identifier = data["identifier"].split("-")[0]
+        instance_groups[identifier]["identifier"] = identifier
+        instance_groups[identifier]["sentence"] = data["sentence"]
+        instance_groups[identifier]["worlds"].append(data["structured_rep"])
+        instance_groups[identifier]["labels"].append(data["label"])
+
+    output_ptr = open(output_file, "w")
```
Comment (matt-gardner) [inline scripts/nlvr/group_nlvr_worlds.py]: It's better in python to do

```python
with open(output_file, 'w') as output:
    json.dump(instance_group, output)
    output.write('\n')
```

File: tests/fixtures/encoder_decoder/nlvr_semantic_parser/grouped_experiment.json [hunk 5]
```diff
@@ -6,8 +6,8 @@
   "vocabulary": {
```
Comment (matt-gardner) [inline tests/fixtures/encoder_decoder/nlvr_semantic_parser/grouped_experiment.json]: For ease of use with our predictor scripts (see `scripts/train_fixtures.py` on the master branch), it's actually helpful to have this stay as `some/path/experiment.json`.

File: tests/fixtures/encoder_decoder/nlvr_semantic_parser/ungrouped_experiment.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
+{
```
Comment (matt-gardner) [inline tests/fixtures/encoder_decoder/nlvr_semantic_parser/ungrouped_experiment.json]: It's fine to have other names in here for other configurations, as long as there is one that's just `experiment.json`, and that's the one you want to test with a `Predictor`.

File: tests/models/encoder_decoders/nlvr/nlvr_semantic_parser_test.py [hunk 7]
```diff
@@ -48,3 +48,13 @@ def test_forward_with_epoch_num_changes_cost_weight(self):
         # The config file has ``wait_num_epochs`` set to 0, so the model starts decreasing the const
         # weight at epoch 0 itself.
         assert_almost_equal(cost_weights, [0.72, 0.648, 0.5832, 0.52488])
+
+
+class NlvrSemanticParserGroupedTest(ModelTestCase):
```
Comment (matt-gardner) [inline tests/models/encoder_decoders/nlvr/nlvr_semantic_parser_test.py]: You don't need another class for this.  See here: https://github.com/allenai/allennlp/blob/31ff177ccca06f044add05cda31f739b8ed7b3fc/tests/models/encoder_decoders/wikitables/wikitables_semantic_parser_test.py#L19-L24

File: allennlp/data/dataset_readers/nlvr.py (hunk 1 above)
Comment (pdasigi) [reply allennlp/data/dataset_readers/nlvr.py]: Fixed.

File: allennlp/data/dataset_readers/nlvr.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/data/dataset_readers/nlvr.py]: Done.

File: allennlp/data/dataset_readers/nlvr.py (hunk 3 above)
Comment (pdasigi) [reply allennlp/data/dataset_readers/nlvr.py]: Done.

File: scripts/nlvr/group_nlvr_worlds.py (hunk 4 above)
Comment (pdasigi) [reply scripts/nlvr/group_nlvr_worlds.py]: Done, thanks!

File: tests/fixtures/encoder_decoder/nlvr_semantic_parser/grouped_experiment.json (hunk 5 above)
Comment (pdasigi) [reply tests/fixtures/encoder_decoder/nlvr_semantic_parser/grouped_experiment.json]: Good to know, done.

File: tests/models/encoder_decoders/nlvr/nlvr_semantic_parser_test.py (hunk 7 above)
Comment (pdasigi) [reply tests/models/encoder_decoders/nlvr/nlvr_semantic_parser_test.py]: Done. Thanks!
